### ISSUE TITLE
JobMemory AI - HP情報取得機能 追加実装指示書

### ISSUE THREAD

**Issue by shimamurayaruzo** [shimamurayaruzo/jobmemory-ai#7]
# JobMemory AI - HP情報取得機能 追加実装指示書

## 概要
既存の求人生成フォームに「会社HPから情報を自動取得する機能」を追加する。
URLを入力して「情報を取得」ボタンを押すと、AIがWebページを解析し、
会社の強み・事業内容などを自動で入力欄に反映する。

## 重要な制約
- simple-job-board/ フォルダは編集禁止（参照のみ）
- 既存の手動入力機能はそのまま残す
- HP取得は「任意」機能（使わなくても求人生成可能）

---

### 追加する入力欄（既存フォームの職種名の下に配置）

```html
<!-- HP情報取得セクション -->
<div class="jmai-hp-section">
    <h3>会社HPから情報を取得（任意）</h3>
    <p class="description">会社概要・採用ページのURLがおすすめです</p>
    
    <div class="jmai-hp-input-row">
        <input type="url" 
               id="jmai-company-url" 
               placeholder="https://example.co.jp/about"
               class="regular-text">
        <button type="button" 
                id="jmai-fetch-hp" 
                class="button button-secondary">
            情報を取得
        </button>
        <span id="jmai-fetch-spinner" class="spinner"></span>
    </div>
    
    <div id="jmai-fetch-message" class="jmai-message"></div>
</div>
Copy
既存の手動入力欄
変更なし。ただしHP取得成功時に以下のフィールドへ自動入力:

会社の強み・魅力 (jmai-company-strength)
事業内容 (jmai-business-description) ※新規追加
職場環境・カルチャー (jmai-work-environment)
新規追加する入力欄
Copy<tr>
    <th><label for="jmai-business-description">事業内容</label></th>
    <td>
        <textarea id="jmai-business-description" 
                  rows="3" 
                  class="large-text"
                  placeholder="例: AI/DXソリューションの開発・提供"></textarea>
    </td>
</tr>
2. JavaScript実装 (assets/admin.js に追加)
Copy// HP情報取得ボタンのクリックイベント
jQuery(document).ready(function($) {
    
    $('#jmai-fetch-hp').on('click', function() {
        var url = $('#jmai-company-url').val().trim();
        
        // バリデーション
        if (!url) {
            $('#jmai-fetch-message')
                .removeClass('success')
                .addClass('error')
                .text('URLを入力してください');
            return;
        }
        
        if (!isValidUrl(url)) {
            $('#jmai-fetch-message')
                .removeClass('success')
                .addClass('error')
                .text('正しいURLを入力してください');
            return;
        }
        
        // ローディング表示
        $('#jmai-fetch-spinner').addClass('is-active');
        $('#jmai-fetch-hp').prop('disabled', true);
        $('#jmai-fetch-message').text('HPから情報を取得中...（10〜20秒）');
        
        // AJAX送信
        $.ajax({
            url: jmai_ajax.ajax_url,
            type: 'POST',
            data: {
                action: 'jmai_fetch_company_info',
                nonce: jmai_ajax.nonce,
                url: url
            },
            success: function(response) {
                $('#jmai-fetch-spinner').removeClass('is-active');
                $('#jmai-fetch-hp').prop('disabled', false);
                
                if (response.success) {
                    // 取得した情報を入力欄に反映
                    var data = response.data;
                    
                    if (data.company_strength) {
                        $('#jmai-company-strength').val(data.company_strength);
                    }
                    if (data.business_description) {
                        $('#jmai-business-description').val(data.business_description);
                    }
                    if (data.work_environment) {
                        $('#jmai-work-environment').val(data.work_environment);
                    }
                    
                    // 成功メッセージ
                    var message = '情報を取得しました。内容を確認・編集してください。';
                    if (data.warning) {
                        message += '\n注意: ' + data.warning;
                    }
                    $('#jmai-fetch-message')
                        .removeClass('error')
                        .addClass('success')
                        .text(message);
                        
                } else {
                    $('#jmai-fetch-message')
                        .removeClass('success')
                        .addClass('error')
                        .text('エラー: ' + response.data.message);
                }
            },
            error: function() {
                $('#jmai-fetch-spinner').removeClass('is-active');
                $('#jmai-fetch-hp').prop('disabled', false);
                $('#jmai-fetch-message')
                    .removeClass('success')
                    .addClass('error')
                    .text('通信エラーが発生しました');
            }
        });
    });
    
    // URL形式チェック
    function isValidUrl(string) {
        try {
            new URL(string);
            return true;
        } catch (_) {
            return false;
        }
    }
});
Copy
3. PHP実装 (includes/class-admin.php に追加)
AJAXアクション登録（コンストラクタ内に追加）
Copyadd_action('wp_ajax_jmai_fetch_company_info', [$this, 'ajax_fetch_company_info']);
HP情報取得メソッド
Copy/**
 * 会社HPから情報を取得してAIで解析
 */
public function ajax_fetch_company_info() {
    // nonceチェック
    if (!check_ajax_referer('jmai_nonce', 'nonce', false)) {
        wp_send_json_error(['message' => '認証エラー']);
    }
    
    $url = isset($_POST['url']) ? esc_url_raw($_POST['url']) : '';
    
    if (empty($url)) {
        wp_send_json_error(['message' => 'URLが入力されていません']);
    }
    
    // URLからHTMLを取得
    $html = $this->fetch_webpage($url);
    
    if (is_wp_error($html)) {
        wp_send_json_error(['message' => $html->get_error_message()]);
    }
    
    // HTMLからテキストを抽出
    $text = $this->extract_text_from_html($html);
    
    if (mb_strlen($text) < 100) {
        wp_send_json_error(['message' => 'ページから十分な情報を取得できませんでした']);
    }
    
    // AIで企業情報を抽出
    $ai_client = new JMAI_AI_Client();
    $extracted = $ai_client->extract_company_info($text);
    
    if (is_wp_error($extracted)) {
        wp_send_json_error(['message' => $extracted->get_error_message()]);
    }
    
    // 警告メッセージ（情報が少ない場合）
    $warning = null;
    if (mb_strlen($text) < 500) {
        $warning = '取得した情報が少なめです。会社概要ページのURLを試すとより詳しい情報が取得できます。';
    }
    
    wp_send_json_success([
        'company_strength' => $extracted['company_strength'] ?? '',
        'business_description' => $extracted['business_description'] ?? '',
        'work_environment' => $extracted['work_environment'] ?? '',
        'warning' => $warning
    ]);
}

/**
 * WebページのHTMLを取得
 */
private function fetch_webpage($url) {
    $args = [
        'timeout' => 30,
        'user-agent' => 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
        'sslverify' => false
    ];
    
    $response = wp_remote_get($url, $args);
    
    if (is_wp_error($response)) {
        return new WP_Error('fetch_error', 'ページを取得できませんでした: ' . $response->get_error_message());
    }
    
    $status_code = wp_remote_retrieve_response_code($response);
    if ($status_code !== 200) {
        return new WP_Error('fetch_error', 'ページを取得できませんでした（ステータス: ' . $status_code . '）');
    }
    
    return wp_remote_retrieve_body($response);
}

/**
 * HTMLからテキストを抽出
 */
private function extract_text_from_html($html) {
    // scriptとstyleタグを除去
    $html = preg_replace('/<script\b[^>]*>(.*?)<\/script>/is', '', $html);
    $html = preg_replace('/<style\b[^>]*>(.*?)<\/style>/is', '', $html);
    
    // HTMLタグを除去
    $text = strip_tags($html);
    
    // 余分な空白・改行を整理
    $text = preg_replace('/\s+/', ' ', $text);
    $text = trim($text);
    
    // 長すぎる場合は切り詰め（トークン節約）
    if (mb_strlen($text) > 5000) {
        $text = mb_substr($text, 0, 5000);
    }
    
    return $text;
}
Copy
4. AIクライアント実装 (includes/class-ai-client.php に追加)
Copy/**
 * 企業情報をAIで抽出
 */
public function extract_company_info($webpage_text) {
    $prompt = <<<PROMPT
以下のWebページから抽出したテキストを分析し、求人作成に必要な企業情報を抽出してください。

【抽出項目】（JSON形式で出力）
{
  "company_strength": "会社の強み・特徴・他社との差別化ポイント（2〜3文）",
  "business_description": "事業内容・提供サービス（2〜3文）",
  "work_environment": "職場環境・企業文化・働き方の特徴（2〜3文）"
}

【注意事項】
- 情報が見つからない項目は空文字 "" にしてください
- 推測や創作はせず、テキストに記載されている情報のみ抽出してください
- 日本語で出力してください
- JSON以外の文字は出力しないでください

【Webページテキスト】
{$webpage_text}
PROMPT;

    $response = $this->call_api([
        [
            'role' => 'system',
            'content' => 'あなたは企業情報を抽出するアシスタントです。指定されたJSON形式でのみ回答してください。'
        ],
        [
            'role' => 'user',
            'content' => $prompt
        ]
    ], 1000, 0.3);
    
    if (is_wp_error($response)) {
        return $response;
    }
    
    // JSONパース
    $content = $response['choices'][0]['message']['content'] ?? '';
    
    // JSON部分を抽出（余分なテキストがある場合に対応）
    if (preg_match('/\{[\s\S]*\}/', $content, $matches)) {
        $content = $matches[0];
    }
    
    $data = json_decode($content, true);
    
    if (json_last_error() !== JSON_ERROR_NONE) {
        return new WP_Error('parse_error', 'AIの応答を解析できませんでした');
    }
    
    return $data;
}
Copy
5. CSS追加 (assets/admin.css に追加)
Copy/* HP情報取得セクション */
.jmai-hp-section {
    background: #f9f9f9;
    border: 1px solid #ddd;
    border-radius: 4px;
    padding: 15px;
    margin-bottom: 20px;
}

.jmai-hp-section h3 {
    margin-top: 0;
    margin-bottom: 5px;
    font-size: 14px;
}

.jmai-hp-section .description {
    color: #666;
    font-size: 12px;
    margin-bottom: 10px;
}

.jmai-hp-input-row {
    display: flex;
    align-items: center;
    gap: 10px;
}

.jmai-hp-input-row input[type="url"] {
    flex: 1;
    max-width: 400px;
}

.jmai-message {
    margin-top: 10px;
    padding: 8px 12px;
    border-radius: 4px;
    font-size: 13px;
}

.jmai-message.success {
    background: #d4edda;
    color: #155724;
    border: 1px solid #c3e6cb;
}

.jmai-message.error {
    background: #f8d7da;
    color: #721c24;
    border: 1px solid #f5c6cb;
}

.jmai-message:empty {
    display: none;
}
Copy
6. 求人生成プロンプトの更新
既存の求人生成プロンプトに「事業内容」を追加:

Copy// class-ai-client.php の generate_job_post メソッド内

$prompt = <<<PROMPT
あなたはGAIS会員企業向けの求人文を作成するAIです。

【GAIS Memory】
{$memory}

【入力情報】
- 職種名: {$job_title}
- 募集背景: {$recruitment_background}
- 事業内容: {$business_description}        // ← 追加
- 会社の強み・魅力: {$company_strength}
- 職場環境・カルチャー: {$work_environment}
- 給与・待遇: {$salary_benefits}
- 求める人物像: {$ideal_candidate}

// ... 以下既存のプロンプト
PROMPT;
7. テスト手順
正常系テスト

会社概要ページのURL入力 → 情報取得成功 → 入力欄に反映
例: https://gais.jp/ でテスト
異常系テスト

空のURL → エラーメッセージ表示
無効なURL → エラーメッセージ表示
存在しないURL → エラーメッセージ表示
情報が少ないページ → 警告メッセージ + 取得できた情報は反映
統合テスト

HP取得 → 情報反映 → 手動で編集 → 求人生成
HP取得せず手動入力のみ → 求人生成（既存機能の確認）
8. 完成チェックリスト
 URL入力欄が表示される
 「情報を取得」ボタンが動作する
 ローディング表示が出る
 取得成功時、入力欄に自動反映される
 取得失敗時、エラーメッセージが表示される
 自動入力後に手動編集できる
 HP取得なしでも従来通り求人生成できる
 事業内容が求人文に反映される

---

**Comment by shimamurayaruzo** [comment]
PR #8 (fix/issue-7) で対応済みです。
